### PR TITLE
Overhaul Movement and Add Shifts

### PR DESCRIPTION
**Pull request by BlairArchibald**
This PR add opening times to places, shifts to people, and overhauls the movement of people through places.

In particular movement is a 2 step approach - first we move everyone to a temporary buffer, then we commit this after everyone has moved. This makes it much easier to avoid ordering effects (e.g. you can no longer move home and then go to work in a single timestep).

### CODE REVIEW

File: src/main/java/uk/co/ramp/covid/simulation/place/Household.java [hunk 1]
```diff
@@ -178,4 +186,106 @@ public void cycleMovements(int day, int hour, boolean lockdown) {
...
+            visitProb = visitProb * 0.5;
+        }
+
+        if (RNG.get().nextUniform(0, 1) < visitProb) {
+            Shop s = places.getRandomShop();
+            if (s == null) {
+                return;
+            }
+            // We go shopping as a family
+            if (s.isVisitorOpenNextHour(day, hour)) {
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/place/Household.java:238]: Do all family members always go shopping together? Do we need to some way to decide how many members of a family go on each shopping trip?

File: src/main/java/uk/co/ramp/covid/simulation/place/Household.java [hunk 2]
```diff
@@ -178,4 +186,106 @@ public void cycleMovements(int day, int hour, boolean lockdown) {
...
+
+    private void moveRestaurant(int day, int hour) {
+        List<Person> left = new ArrayList();
+
+        double visitProb = PopulationParameters.get().getpGoRestaurant();
+        //TODO: Handle lockdown probabilities
+
+        if (RNG.get().nextUniform(0, 1) < visitProb) {
+            Restaurant r = places.getRandomRestaurant();
+            // We go to restaurants as a family
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/place/Household.java:260]: Same question as for shops above

File: src/main/java/uk/co/ramp/covid/simulation/place/Restaurant.java [hunk 3]
```diff
@@ -15,22 +15,25 @@ public Restaurant() {
     public Restaurant(Size s) {
         super(s);
         transProb = PopulationParameters.get().getpBaseTrans() * PopulationParameters.get().getpRestaurantTrans();
-        startDay = 1;
-        endDay = 7;
-        startTime = 10;
-        endTime = 22;
+        times.setOpen(10);
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/place/Restaurant.java:18]: Should we make restaurant open and close times variable. Not all restaurants open at 10am

File: src/main/java/uk/co/ramp/covid/simulation/place/Household.java (hunk 1 above)
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/place/Household.java:238]: Paul suggests (for now at least) they all go and return to/from restaurants/shops/neighbours together.

File: src/main/java/uk/co/ramp/covid/simulation/place/Restaurant.java (hunk 3 above)
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/place/Restaurant.java:18]: Paul has suggested a set of times (based on shop sizes). These are coming in a future commit!

**Comment by BlairArchibald** [comment]
We possibly want to add some more testing of the movement at some stage (although there is some basic sanity checks added already). However I think to make the review more manageable, and to let us start working on the other issues that rely on movement, I think it might be best to add these later.

**Pull request merged**